Week8

sinorga opened the pull request.

sinorga committed:
Assignment: Create subscriptions

sinorga committed:
store strip customer id in user model

sinorga committed:
Assignment: Admin views payments, add payment model and handle webhook of stripe

sinorga committed:
Assignment: Admin views payments(sol)
1. remove PaymentDecorator and replace with number_to_currency in view.
2. move updating stripe_id of user to user_registrator
3. rename stripe webhooks url
4. add request spec for strip event post
5. add feature test for admin view payments

sinorga committed:
Assignment: Handle payment failures

tomtomecek commented on app/controllers/sessions_controller.rb:
```diff
@@ -6,8 +6,13 @@ def new
-      redirect_to home_path, flash: {success: "You are signed in, enjoy!"}
+      if user.active?
+        session[:user_id] = user.id
+        redirect_to home_path, flash: {success: "You are signed in, enjoy!"}
```
Instead of doing this - check [add_flash_types](http://api.rubyonrails.org/classes/ActionController/Flash/ClassMethods.html#method-i-add_flash_types)

sinorga replied:
Oh! Cool, thanks

tomtomecek commented on app/models/stripe_wrapper.rb:
```diff
@@ -1,22 +1,65 @@
 module StripeWrapper
+  def self.set_api_key(key)
+    Stripe.api_key = key
```
I would move this to its own initializer, so the key is set when your app starts.

sinorga replied:
Is it mean people usually won't use wrapper in initializer?

tomtomecek replied:
Well I have noticed you have set it in initializer and have used this method with wrapper. It is OK, however I see it like this first time.

sinorga replied:
I got it, thanks :)

tomtomecek commented on app/services/user_registrator.rb:
```diff
@@ -38,11 +38,11 @@ def register!(stripe_token, invite_token)
+    customer = StripeWrapper::Customer.create(
       :source => stripe_token,
-      :description => "payment of #{user.email}"
+      :user => user
```
Why do you use old ruby hash syntax here? You have fixed it in your StripeWrapper.

sinorga replied:
Yes, I missed this part :p I would correct it.

tomtomecek commented on config/initializers/stripe.rb:
```diff
@@ -1 +1,6 @@
+StripeWrapper.set_api_key(ENV['STRIPE_SECRET_KEY'])
+
+StripeEvent.configure do |events|
+  events.subscribe 'charge.succeeded', StripeWrapper::EventHandler::ChargeSucceeded.new
```
It feels a little bit wrong to create a new class for a new state. Succeeded and Failed.

tomtomecek commented on spec/requests/charge_payment_spec.rb:
```diff
@@ -0,0 +1,79 @@
+    let!(:alice) { Fabricate(:user, stripe_id: "cus_6AigMqcxfS02rf") }
+    before do
+      stub_event
+      post "/stripe_events", JSON.parse(event_data)
```
Nice, I like you have refactored it to its own file :)

sinorga committed:
refactor:
1. using add_flash_types
2. replace old sytle hash
3. combine stripe event handler
4. set stripe api key directly in initializer.
